File: apps/radio-station/server/services/export-txt.ts
import type {PlaylistEntryRow, Weights} from '../types.js';

export interface TxtInput {
    stationName: string;
    seed: string;
    weights: Weights;
    entries: PlaylistEntryRow[];
    timeZone: string;
}

function ymd(ms: number, timeZone: string): string {
    return new Intl.DateTimeFormat('en-CA', {
        timeZone,
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
    }).format(new Date(ms));
}

function hm(ms: number, timeZone: string): string {
    return new Intl.DateTimeFormat('en-GB', {
        timeZone,
        hour: '2-digit',
        minute: '2-digit',
        hourCycle: 'h23',
    }).format(new Date(ms));
}

function weekEndDate(entries: PlaylistEntryRow[], timeZone: string): string {
    const first = entries[0];
    if (!first) return '';
    return ymd(first.startsAt + 7 * 24 * 60 * 60 * 1_000, timeZone);
}

export function formatPlaylistTxt(input: TxtInput): string {
    const first = input.entries[0];
    const start = first ? ymd(first.startsAt, input.timeZone) : '';
    const end = weekEndDate(input.entries, input.timeZone);
    const w = input.weights;
    const header = [
        `${input.stationName} — ${start} to ${end}`,
        `seed: ${input.seed}`,
        `hitGravity=${w.hitGravity} goldLeak=${w.goldLeak} temperature=${w.temperature} separation=${w.separation} powerOrbitMin=${w.powerOrbitMin}`,
        '',
    ].join('\n');
    const rows = input.entries.map(
        (entry) => `${ymd(entry.startsAt, input.timeZone)} ${hm(entry.startsAt, input.timeZone)}  ${entry.artist} — ${entry.title}`,
    );
    return `${header}${rows.join('\n')}\n`;
}

File: apps/radio-station/api/src/main/kotlin/radioapi/domain/ExportTxt.kt
package radioapi.domain

import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

data class TxtEntry(val startsAt: Long, val artist: String, val title: String)

data class TxtInput(
    val stationName: String,
    val seed: String,
    val weights: Weights,
    val entries: List<TxtEntry>,
    val timeZone: String,
)

fun formatPlaylistTxt(input: TxtInput): String {
    val zone = ZoneId.of(input.timeZone)
    val first = input.entries.firstOrNull()
    val start = if (first == null) "" else ymd(first.startsAt, zone)
    val end = if (first == null) "" else ymd(first.startsAt + WEEK_MS, zone)
    val header = txtHeader(input, start, end)
    val rows = input.entries.map { entry ->
        "${ymd(entry.startsAt, zone)} ${hm(entry.startsAt, zone)}  ${entry.artist} — ${entry.title}"
    }
    return header + rows.joinToString("\n") + "\n"
}

private fun txtHeader(input: TxtInput, start: String, end: String): String {
    val weights = input.weights
    return "${input.stationName} — $start to $end\n" +
        "seed: ${input.seed}\n" +
        "hitGravity=${weights.hitGravity} goldLeak=${weights.goldLeak} " +
        "temperature=${weights.temperature} separation=${weights.separation} " +
        "powerOrbitMin=${weights.powerOrbitMin}\n\n"
}

private fun ymd(ms: Long, zone: ZoneId): String =
    DateTimeFormatter.ofPattern("yyyy-MM-dd").withZone(zone).format(Instant.ofEpochMilli(ms))

private fun hm(ms: Long, zone: ZoneId): String =
    DateTimeFormatter.ofPattern("HH:mm").withZone(zone).format(Instant.ofEpochMilli(ms))
